test/replication-luatest/gh_7318_box_wait_limbo_acked_test.lua

```diff
@@ -0,0 +1,525 @@
+end
+
+g.before_each(function(cg)
+    t.tarantool.skip_if_not_debug()
+
+    cg.cluster = cluster:new({})
+    cg.master = cg.cluster:build_and_add_server({
+        alias = 'master',
+        box_cfg = {
+            replication = {
+                server.build_listen_uri('master', cg.cluster.id),
+                server.build_listen_uri('replica_proxy'),
+            },
+            election_mode = 'candidate',
+            replication_synchro_quorum = 2,
+            replication_synchro_timeout = 100000,
+            replication_timeout = 0.1,
+            election_fencing_mode='off',
+        }
+    })
+    cg.replica = cg.cluster:build_and_add_server({
+        alias = 'replica',
+        box_cfg = {
+            replication = {
+                server.build_listen_uri('replica', cg.cluster.id),
+                server.build_listen_uri('master_proxy'),
+            },
+            election_mode = 'off',
+            replication_synchro_quorum = 2,
+            replication_synchro_timeout = 100000,
+            replication_timeout = 0.1,
+            election_fencing_mode='off',
+        }
+    })
+    cg.master_proxy = proxy:new({
+        client_socket_path = server.build_listen_uri('master_proxy'),
+        server_socket_path = server.build_listen_uri('master', cg.cluster.id),
+    })
+    t.assert(cg.master_proxy:start({force = true}))
+    cg.replica_proxy = proxy:new({
+        client_socket_path = server.build_listen_uri('replica_proxy'),
+        server_socket_path = server.build_listen_uri('replica', cg.cluster.id),
+    })
+    t.assert(cg.replica_proxy:start({force = true}))
+    cg.cluster:start()
+    cg.master:wait_until_election_leader_found()
+    cg.replica:wait_until_election_leader_found()
+
+    cg.master:exec(function()
+        box.schema.space.create('test', {is_sync = true})
+        box.space.test:create_index('pk')
+    end)
+    wait_pair_sync(cg.replica, cg.master)
+end)
+
+g.after_each(function(cg)
+    cg.cluster:drop()
+    cg.master_proxy:stop()
+    cg.replica_proxy:stop()
+end)
+
+g.test_wal_sync_fails = function(cg)
+    local f = cg.replica:exec(function()
+        box.error.injection.set('ERRINJ_WAL_DELAY_COUNTDOWN', 0)
+        local f = require('fiber').create(function()
+            return pcall(box.ctl.promote)
+        end)
+        f:set_joinable(true)
+        return f:id()
+    end)
+    server_wait_wal_is_blocked(cg.replica)
+    cg.master:exec(function()
+        require('fiber').create(function() box.space.test:insert{1} end)
+    end)
+    server_wait_synchro_queue_len_is_equal(cg.replica, 1)
+    local csw = get_fiber_csw(cg.replica, f)
+    cg.replica:exec(function()
+        box.error.injection.set('ERRINJ_WAL_SYNC', true)
+        box.error.injection.set('ERRINJ_WAL_DELAY_COUNTDOWN', 0)
+        box.error.injection.set('ERRINJ_WAL_DELAY', false)
+    end)
+    -- We need to be sure we entered the wal_sync call.
+    server_wait_fiber_csw_ge_than(cg.replica, f, csw + 1)
```

Comment:
This is a very hacky way to rely on the number of yields. Here and in all the other places. `server_wait_fiber_csw_ge_than()` shouldn't really exist unless you are testing something exactly about specific number of yields.

In this particular case it seems you could just wait on the wal becoming blocked, no?

Another approach would be to move `ERRINJ_WAL_SYNC_DELAY` before `cbus_call()` in `wal_sync()` and make it an integer. When 0, it is not used. When you set it to 1 in a test, then it is turned into 2 by `wal_sync()` + it starts waiting. Then in your tests you can do `box.error.injection.set('ERRINJ_WAL_SYNC_DELAY', 1)` and then wait until `box.error.injection.get('ERRINJ_WAL_SYNC_DELAY') == 2`.

Reply:
Yes you are right, here I can just use server_wait_wal_is_blocked. But in some tests, I use `server_wait_fiber_csw_ge_than()` to wait until it reaches `box_wait_quorum()`. And I don't know of a way to make sure that out fiber `f` currently staying in `box_wait_quorum()` other than adding another one errinj. I've left a comment below where this happens.

Reply:
You can add another errinj similarly to how `ERRINJ_WAL_WRITE_COUNT` works. It doesn't have to an error. It can be a counter which you can fetch in the test. For example, number of fibers inside of `box_wait_quorum()` right now. It will be fully legal and won't introduce any unrealistic errors (or any errors at all).

Reply:
I did it.
```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 8940153df..4aa8f75fc 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -2404,6 +2404,10 @@ static int
 box_wait_quorum(uint32_t lead_id, int64_t target_lsn, int quorum,
                double timeout)
 {
+#ifndef NDEBUG
+    ++errinj(ERRINJ_WAIT_QUORUM_COUNT, ERRINJ_INT)->iparam;
+#endif
+
        struct box_quorum_trigger t;
        memset(&t, 0, sizeof(t));
        vclock_create(&t.vclock);
diff --git a/src/lib/core/errinj.h b/src/lib/core/errinj.h
index 4e1ec4ea2..61d5bf8f4 100644
--- a/src/lib/core/errinj.h
+++ b/src/lib/core/errinj.h
@@ -170,6 +170,7 @@ struct errinj {
        _(ERRINJ_VY_STMT_ALLOC, ERRINJ_INT, {.iparam = -1})\
        _(ERRINJ_VY_TASK_COMPLETE, ERRINJ_BOOL, {.bparam = false}) \
        _(ERRINJ_VY_WRITE_ITERATOR_START_FAIL, ERRINJ_BOOL, {.bparam = false})\
+       _(ERRINJ_WAIT_QUORUM_COUNT, ERRINJ_INT, {.iparam = 0}) \
        _(ERRINJ_WAL_BREAK_LSN, ERRINJ_INT, {.iparam = -1}) \
        _(ERRINJ_WAL_DELAY, ERRINJ_BOOL, {.bparam = false}) \
        _(ERRINJ_WAL_DELAY_COUNTDOWN, ERRINJ_INT, {.iparam = -1}) \
diff --git a/test/replication-luatest/gh_7318_box_wait_limbo_acked_test.lua b/test/replication-luatest/gh_7318_box_wait_limbo_acked_test.lua
index 431797957..143efe57d 100644
--- a/test/replication-luatest/gh_7318_box_wait_limbo_acked_test.lua
+++ b/test/replication-luatest/gh_7318_box_wait_limbo_acked_test.lua
@@ -56,21 +56,18 @@ local function server_becomes_the_leader_again(server)
     end, {prev_cfg})
 end
 
-local function get_fiber_csw(server, fiber_id)
-    return server:exec(function(fiber_id)
-        return require('fiber').find(fiber_id):info().csw
-    end, {fiber_id})
+local function get_wait_quorum_count(server)
+    return server:exec(function()
+        return box.error.injection.get('ERRINJ_WAIT_QUORUM_COUNT')
+    end)
 end
 
-local function server_wait_fiber_csw_ge_than(server, fiber_id, threshold)
-    server:exec(function(fiber_id, threshold, wait_timeout)
-        t.helpers.retrying({timeout = wait_timeout},
-            function(fiber_id, threshold)
-                local f = require('fiber').find(fiber_id)
-                t.assert_ge(f:info().csw, threshold)
-            end,
-        fiber_id, threshold)
-    end, {fiber_id, threshold, wait_timeout})
+local function server_wait_wait_quorum_count_ge_than(server, threshold)
+    server:exec(function(threshold, wait_timeout)
+        t.helpers.retrying({timeout = wait_timeout}, function(threshold)
+            t.assert_ge(box.error.injection.get('ERRINJ_WAIT_QUORUM_COUNT'), threshold)
+        end, threshold)
+    end, {threshold, wait_timeout})
 end
 
 local function server_wait_synchro_queue_is_busy(server)
@@ -367,7 +364,7 @@ g.test_update_greatest_term_while_wait_quorum = function(cg)
         box.error.injection.set('ERRINJ_WAL_DELAY', false)
     end)
     server_wait_wal_is_blocked(cg.replica)
-    local csw = get_fiber_csw(cg.replica, f)
+    local wait_quorum_count = get_wait_quorum_count(cg.replica)
     local ff = require('fiber').create(function()
         cg.replica:exec(function(f)
             box.error.injection.set('ERRINJ_WAL_DELAY', false)
@@ -381,7 +378,7 @@ g.test_update_greatest_term_while_wait_quorum = function(cg)
     end)
     ff:set_joinable(true)
     -- We need to be sure we entered the 'box_wait_quorum' call.
-    server_wait_fiber_csw_ge_than(cg.replica, f, csw + 1)
+    server_wait_wait_quorum_count_ge_than(cg.replica, wait_quorum_count + 1)
     server_becomes_the_leader_again(cg.master)
     server_wait_promote_greatest_term_ge_than(cg.replica, 3)
     -- Replica collects quorum.
@@ -434,7 +431,7 @@ g.test_txn_limbo_is_empty_while_wait_quorum = function(cg)
     cg.master:exec(function()
         box.cfg{replication_synchro_quorum=2}
     end)
-    local csw = get_fiber_csw(cg.replica, f)
+    local wait_quorum_count = get_wait_quorum_count(cg.replica)
     local ff = require('fiber').create(function()
         cg.replica:exec(function(f)
             box.error.injection.set('ERRINJ_WAL_DELAY', false)
@@ -444,7 +441,7 @@ g.test_txn_limbo_is_empty_while_wait_quorum = function(cg)
         end, {f})
     end)
     ff:set_joinable(true)
-    server_wait_fiber_csw_ge_than(cg.replica, f, csw + 1)
+    server_wait_wait_quorum_count_ge_than(cg.replica, wait_quorum_count + 1)
     -- Now the master will send 'CONFIRM'.
     cg.master_proxy:resume()
     server_wait_synchro_queue_len_is_equal(cg.replica, 0)
@@ -545,7 +542,7 @@ g.test_new_synchronous_transactions_appeared_while_wait_quorum = function(cg)
         box.error.injection.set('ERRINJ_WAL_DELAY', false)
     end)
     server_wait_wal_is_blocked(cg.replica)
-    local csw = get_fiber_csw(cg.replica, f)
+    local wait_quorum_count = get_wait_quorum_count(cg.replica)
     local ff = require('fiber').create(function()
         cg.replica:exec(function(f)
             box.error.injection.set('ERRINJ_WAL_DELAY', false)
@@ -557,7 +554,7 @@ g.test_new_synchronous_transactions_appeared_while_wait_quorum = function(cg)
         end, {f})
     end)
     ff:set_joinable(true)
-    server_wait_fiber_csw_ge_than(cg.replica, f, csw + 1)
+    server_wait_wait_quorum_count_ge_than(cg.replica, wait_quorum_count + 1)
     -- Now the master will send the second insert transaction.
     cg.master_proxy:resume()
     server_wait_synchro_queue_len_is_equal(cg.replica, 2)
```

Reply:
Resolving since @Gerold103 approved the PR.